apps/api/src/app/rate-limiting/guards/throttler.guard.ts

```diff
@@ -0,0 +1,182 @@
+    // Return early if the current user agent should be ignored.
+    if (Array.isArray(ignoreUserAgents)) {
+      for (const pattern of ignoreUserAgents) {
+        if (pattern.test(req.headers[RateLimitHeaderKeysEnum.USER_AGENT.toLowerCase()])) {
+          return true;
+        }
+      }
+    }
+
+    const handler = context.getHandler();
+    const classRef = context.getClass();
+    const apiRateLimitCategory =
+      this.reflector.getAllAndOverride(ThrottlerCategory, [handler, classRef]) || defaultApiRateLimitCategory;
+    const apiRateLimitCost =
+      this.reflector.getAllAndOverride(ThrottlerCost, [handler, classRef]) || defaultApiRateLimitCost;
+
+    const { organizationId, environmentId } = this.getReqUser(context);
+
+    const { success, limit, remaining, reset, windowDuration, burstLimit, algorithm } =
+      await this.evaluateApiRateLimit.execute(
+        EvaluateApiRateLimitCommand.create({
+          organizationId,
+          environmentId,
+          apiRateLimitCategory,
+          apiRateLimitCost,
+        })
+      );
+
+    const secondsToReset = Math.max(Math.ceil((reset - Date.now()) / 1e3), 0);
+
+    res.header(RateLimitHeaderKeysEnum.RATE_LIMIT_REMAINING, remaining);
+    res.header(RateLimitHeaderKeysEnum.RATE_LIMIT_LIMIT, limit);
+    res.header(RateLimitHeaderKeysEnum.RATE_LIMIT_RESET, secondsToReset);
+    res.header(
+      RateLimitHeaderKeysEnum.RATE_LIMIT_POLICY,
+      this.createPolicyHeader(limit, windowDuration, burstLimit, algorithm, apiRateLimitCategory, apiRateLimitCost)
+    );
+
+    if (success) {
+      return true;
+    } else {
+      res.header(RateLimitHeaderKeysEnum.RETRY_AFTER, secondsToReset);
+      throw new ThrottlerException(THROTTLED_EXCEPTION_MESSAGE);
+    }
+  }
+
+  private createPolicyHeader(
+    limit: number,
+    windowDuration: number,
+    burstLimit: number,
+    algorithm: string,
+    apiRateLimitCategory: ApiRateLimitCategoryEnum,
+    apiRateLimitCost: ApiRateLimitCostEnum
+  ): string {
+    const policyMap = {
+      w: windowDuration,
+      burst: burstLimit,
+      comment: `"${algorithm}"`,
+      category: `"${apiRateLimitCategory}"`,
+      cost: `"${apiRateLimitCost}"`,
+    };
+    const policy = Object.entries(policyMap).reduce((acc, [key, value]) => {
+      return `${acc};${key}=${value}`;
+    }, `${limit}`);
+
+    return policy;
+  }
+
+  private isAllowedAuthScheme(context: ExecutionContext): boolean {
+    const req = context.switchToHttp().getRequest();
+    const authorization = req.headers[RateLimitHeaderKeysEnum.AUTHORIZATION.toLowerCase()];
+    if (!authorization) {
+      return false;
+    }
+
+    return ALLOWED_AUTH_SCHEMES.some((scheme) => req.authScheme === scheme);
+  }
+
+  private getReqUser(context: ExecutionContext): IJwtPayload {
+    const req = context.switchToHttp().getRequest();
+    const token = req.headers[RateLimitHeaderKeysEnum.AUTHORIZATION.toLowerCase()].split(' ')[1];
```

Comment:
See https://github.com/novuhq/novu/pull/4910#discussion_r1409001401

Reply:
Right now, we are overwriting the `Authorization: ApiKey: XXXXX` header with an `Authorization: Bearer XXXXX` when a request uses API key auth. So this will always be a JWT token right now.

https://linear.app/novu/issue/NV-3055/🔐-fix-apikey-auth-guard-performance will clean this logic up, and we can adjust the flow here - notably, the `Remove all use of JWT (e.g. jwt.decode)` in the task description will deal with this particular instance.

Reply:
@djabarovgeorge I have prioritized 3055 in the next cycle as this debt.